src/engine/Service.cpp

```diff
@@ -193,30 +195,30 @@ std::optional<std::string> Service::getSiblingValuesClause() const {
     vars += it->first.name() + " ";
     commonColumnIndices.push_back(it->second.columnIndex_);
   }
-  vars.pop_back();
-  if (commonColumnIndices.size() > 1) {
-    vars = "(" + vars + ")";
-  }
+  vars.back() = ')';
+
+  ad_utility::HashSet<std::string> rowSet;
 
   std::string values = " { ";
   for (size_t rowIndex = 0; rowIndex < siblingResult->size(); ++rowIndex) {
-    std::string row;
+    std::string row = "(";
     for (size_t i = 0; i < commonColumnIndices.size(); ++i) {
       const auto& optionalString = ExportQueryExecutionTrees::idToStringAndType(
           siblingTree_->getRootOperation()->getIndex(),
           siblingResult->idTable()(rowIndex, commonColumnIndices[i]),
           siblingResult->localVocab());
 
       if (optionalString.has_value()) {
-        row += optionalString.value().first;
-        if (i < commonColumnIndices.size() - 1) {
-          row += " ";
-        }
+        row += optionalString.value().first + ' ';
       }
     }
-    if (commonColumnIndices.size() > 1) {
-      row = "(" + row + ")";
+    row.back() = ')';
+
+    if (rowSet.contains(row)) {
+      continue;
     }
+
+    rowSet.insert(row);
     values += row + " ";
```

Comment:
For all your `someString += something + something other`
use `absl::StrAppend`. It is more efficient and more idiomatic. (three places in this file)